src/dawn/SIR/ASTNodeIterator.h

```diff
@@ -0,0 +1,308 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_SIR_ASTASTNODEITERATOR_H
+#define DAWN_SIR_ASTASTNODEITERATOR_H
+
+#include "ASTStmt.h"
+#include "dawn/Support/IteratorRange.h"
+#include <list>
+#include <type_traits>
+
+namespace dawn {
+
+/// @brief (Abstract) Pre-order visiting iterator through the Stmts of one or more ASTs.
+/// Specializations should specify on which ASTs to iterate and in which order.
+/// @ingroup sir
+template <typename GeneratorType, bool onlyFirstLevel, typename Enable = void>
+class ASTNodeIterator {
+public:
+  virtual ~ASTNodeIterator() = 0;
+};
+
+#define ASTNodeIterator_Stmt                                                                       \
+  ASTNodeIterator<ASTNode, onlyFirstLevel,                                                         \
+                  typename std::enable_if<std::is_base_of<Stmt, ASTNode>::value>::type>
+#define ASTVoidIterator_Stmt ASTVoidIterator<ASTNode, onlyFirstLevel>
+
+/// @brief Pre-order visiting iterator through the Stmts of an AST. Generated from a root Stmt.
+/// Depending on onlyFirstLevel parameter the visit is/isn't limited to the first level (also
+/// excluding root) of the AST.
+/// @ingroup sir
+template <typename ASTNode, bool onlyFirstLevel>
+class ASTNodeIterator_Stmt {
+  friend class ASTOps;
+  template <typename T, bool B, typename E>
+  friend class ASTNodeIterator;
```

Comment:
I will summarize our discussion around this matter:
The best design choice would be to avoid using friends at all. 
In this particular case however https://github.com/Stagno/dawn/blob/1c931711ace2a31813f302755fff16abb9ced048/src/dawn/SIR/ASTNodeIterator.h#L46-L47 we didn't find (up to now) a better solution since constructing a "top-level" `ASTNodeIterator` requires also constructing its "rest" iterator, whose constructor is protected and must remain so. Same problem solved with the same workaround in this PR (https://github.com/MeteoSwiss-APN/dawn/pull/216).
In the case of `ASTOps`, instead, we could consider embedding `ASTOps` inside `ASTNodeIterator` to avoid friending them.

Regarding the final question: solved by https://github.com/Stagno/dawn/pull/2#discussion_r304832399 ?

Reply:
Update: how to unfriend ASTOps from ASTNodeIterator, ASTNodeIteratorImpl, etc. is still an open issue. Putting it as a nested class inside ASTNodeIterator only gives private/protected accesses to ASTNodeIterator and not to the other classes' methods/members.